src/main/kotlin/com/hedvig/accountservice/termination/TerminationService.kt

```diff
@@ -69,21 +91,107 @@ class TerminationService @Autowired constructor(
+                    this.notificationService.sendFailedChargesTerminationNotification(memberId, requestId)
+                }
+            } catch (exception: Exception) {
+                Sentry.capture(exception)
+                logger.error { "Error when checking if member should be terminated (Exception: $exception)" }
+
+                if ((isActiveContract(contract, memberId))) {
+                    this.notificationService.sendFailedChargeNotConnectedDirectDebitNotification(memberId, requestId)
+                }
+            }
+        }
+    }
+
+    @EventHandler
+    fun on(
+        event: ChargeFailedEvent,
+        @Timestamp timestamp: Instant,
+        replayStatus: ReplayStatus,
+        domainEventMessage: DomainEventMessage<Any>
+    ) {
+        if (!replayStatus.isReplay) {
+            notifyAndMaybeTerminateMember(
+                memberId = event.memberId,
+                requestId = domainEventMessage.identifier,
+                timestamp = timestamp,
+                notification = this.notificationService::sendFailedChargeInsufficientFundsNotification
+            )
+        }
+    }
+
+    @EventHandler
+    fun on(
+        event: ChargeRequestFailedEvent,
+        @Timestamp timestamp: Instant,
+        replayStatus: ReplayStatus,
+        domainEventMessage: DomainEventMessage<Any>
+    ) {
+        if (!replayStatus.isReplay) {
+            notifyAndMaybeTerminateMember(
+                memberId = event.memberId,
+                requestId = domainEventMessage.identifier,
+                timestamp = timestamp,
+                notification = this.notificationService::sendFailedChargeNotConnectedDirectDebitNotification
+            )
+        }
+    }
+
+    private fun isActiveContract(contract: ContractDto, memberId: String): Boolean {
+        val lastDayOfMonth = accountManager.getLatestSubscriptionEntry(memberId)!!.fromDate
+            .plusMonths(1)
+            .withDayOfMonth(1)
+            .minusDays(1)
+
+        if (contract.masterInception == null ||
+            contract.masterInception.isAfter(ChronoLocalDate.from(lastDayOfMonth)) ||
+            contract.isTerminated) {
+            return false
+        }
+
+        return true
+    }
+
+    private fun terminateContractOnLastDayOfMonth(
         memberId: String,
         timestamp: Instant,
-        product: InsuranceDTO,
+        contract: ContractDto,
         numberFailedCharges: Int,
-        personFlags: PersonFlags
+        personFlags: PersonFlags,
+        terminationReason: TerminationReason
     ) {
         val lastDayOfMonth = accountManager.getLatestSubscriptionEntry(memberId)!!.fromDate
             .plusMonths(1)
             .withDayOfMonth(1)
             .minusDays(1)
-        productPricingService.cancelProduct(memberId, product.productId, lastDayOfMonth)
+
+        val terminationRequest = TerminateContractRequest(
+            timestamp.atZone(ZoneId.of("Europe/Stockholm")).toLocalDate(),
```

Comment:
We should terminate on `lastDayOfMonth` not today

Reply:
The function is called `terminateContractOnLastDayOfMonth` for a reason ;)

Reply:
Adding names to call arguments would be a good improvement as well
```suggestion
           terminationDate = timestamp.atZone(ZoneId.of("Europe/Stockholm")).toLocalDate(),
```